CONTRIBUTING.md

```diff
@@ -31,11 +39,171 @@ pgbackrest-dev => Clone pgBackRest repository
+
+## Coding
+
+The following sections provide information on some important concepts needed for coding within pgBackRest.
+
+### Memory Context
+
+Memory is allocated inside contexts and can be long lasting (for objects) or temporary (for functions). In general, use `MEM_CONTEXT_NEW_BEGIN("SomeName")` for objects and `MEM_CONTEXT_TEMP_BEGIN()` for functions. See [memContext.h](https://github.com/pgbackrest/pgbackrest/blob/master/src/common/memContext.h) for more details and the [Coding Example](#coding-example) below.
+
+### Logging
+
+Logging is used in debugging with the built-in macros FUNCTION_LOG_* and FUNCTION_TEST_* which are used to trace parameters passed to/returned from functions. FUNCTION_LOG_* macros are used for production logging whereas FUNCTION_TEST_* macros will be compiled out of production code. For functions where no parameter is valuable for debugging in production, use `FUNCTION_TEST_BEGIN()/FUNCTION_TEST_END()`, else use `FUNCTION_LOG_BEGIN(someLogLevel)/FUNCTION_LOG_END()`. See [debug.h](https://github.com/pgbackrest/pgbackrest/blob/master/src/common/debug.h) for more details and the [Coding Example](#coding-example) below.
+
+Logging is also used for providing information to the user via the LOG_* macros, such as `LOG_INFO("some informational message")` and `LOG_WARN_FMT("no prior backup exists, %s backup has been changed to full", strZ(cfgOptionDisplay(cfgOptType)))` and also via THROW_* macros when throwing an error. See [log.h](https://github.com/pgbackrest/pgbackrest/blob/master/src/common/log.h) and [error.h](https://github.com/pgbackrest/pgbackrest/blob/master/src/common/error.h) for more details and the [Coding Example](#coding-example) below.
+
+### Coding Example
+
+In the hypothetical example below, code comments are not indented as would be appropriate in the code as they are only here to explain the example. Refer to the sections above and [CODING.md](https://github.com/pgbackrest/pgbackrest/blob/master/CODING.md) for an introduction to the details provided here.
+
+#### Example: basic object construction
+```c
+// Declare the publicly accessible variables in a structure with Pub appended to the name
+typedef struct MyObjPub         // First letter upper case
+{
+    MemContext *memContext;     // Pointer to memContext in which this object resides
+    unsigned int myData;        // Contents of the myData variable
+} MyObjPub;
+
+// Declare the object type
+struct MyObj
+{
+    MyObjPub pub;               // Publicly accessible variables must be first and named "pub"
+    const String *name;         // Pointer to lightweight string object - see string.h
+};
+
+// Declare getters and setters inline for the publicly visible variables.
+// Only setters require "Set" appended to the name.
+__attribute__((always_inline)) static inline unsigned int
+myObjMyData(const MyObj *const this)
+{
+    return THIS_PUB(MyObj)->myData;    // Use the built-in THIS_PUB macro
+}
+
+// TYPE and FROMAT macros for function logging
```

Comment:
**TYPE and FORMAT** ?

Reply:
good catch - thanks! Fixed in 269b9f13e715b061cc3d006c62e0b32525af3f3f